presto-cli/src/main/java/com/facebook/presto/cli/QueryPreprocessor.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.cli;
+
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableList;
+import com.google.common.io.CharStreams;
+import com.google.common.util.concurrent.SettableFuture;
+import io.airlift.units.Duration;
+import sun.misc.Signal;
+import sun.misc.SignalHandler;
+
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.facebook.presto.cli.ConsolePrinter.REAL_TERMINAL;
+import static com.google.common.base.Strings.emptyToNull;
+import static com.google.common.base.Throwables.propagateIfPossible;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+public final class QueryPreprocessor
+{
+    public static final String ENV_PREPROCESSOR = "PRESTO_PREPROCESSOR";
+    public static final String ENV_PREPROCESSOR_TIMEOUT = "PRESTO_PREPROCESSOR_TIMEOUT";
+    public static final String ENV_PRESTO_CATALOG = "PRESTO_CATALOG";
+    public static final String ENV_PRESTO_SCHEMA = "PRESTO_SCHEMA";
+    private static final Duration DEFAULT_PREPROCESSOR_TIMEOUT = new Duration(10, SECONDS);
+
+    private static final Signal SIGINT = new Signal("INT");
+    private static final String PREPROCESSING_QUERY_MESSAGE = "Preprocessing query";
+
+    private QueryPreprocessor() {}
+
+    public static String preprocessQuery(Optional<String> catalog, Optional<String> schema, String query)
+            throws QueryPreprocessorException
+    {
+        Duration timeout = DEFAULT_PREPROCESSOR_TIMEOUT;
+        String timeoutEnvironment = System.getenv(ENV_PREPROCESSOR_TIMEOUT);
+        if (emptyToNull(timeoutEnvironment) != null) {
+            timeout = Duration.valueOf(timeoutEnvironment.trim());
+        }
+
+        String preprocessorCommand = System.getenv(ENV_PREPROCESSOR);
+        if (emptyToNull(preprocessorCommand) == null) {
+            return query;
+        }
+        return preprocessQuery(catalog, schema, query, ImmutableList.of("/bin/sh", "-c", preprocessorCommand.trim()), timeout);
```

Comment:
Sorry :)

> When PRESTO_PREPROCESSOR environment variable is set, all commands
are piped through the specified program

"program" doesn't tell this can be "program + arguments", a pipeline or even full-blown shell script in sh. I think hinting at this could be nice in the commit message.